accepted/future-releases/language-versioning/package-config-file-v2.md

```diff
@@ -0,0 +1,134 @@
+# Dart Package Configuration File v2.0
+
+Author: lrn@google.com
+Version: 0.1
+
+Dart currently uses a `.packages` file to configure *package URI resolution* (converting a package URI to a file name). The file uses a `.ini` file-like format with package names as keys and URI references as values. The URI references generated by Pub are always absolute or relative paths, so third-party users of the file tend to see them as paths, not URI.
+
+For the *language versioning* feature, we have proposed extending the format by adding fragments to the URI references and an entry with the empty string as key. This is a breaking change (the empty key was not previously accepted) and even more breaking for tools which assume the values are paths.
+
+Instead of breaking the existing file users, we will instead introduce a *new* file containing the information that we intended to store in the `.packages` file. This fill will use a (more) structured format which allows extending it with further information in the future. We will still generate the `.packages` file for now, allowing users of the file to continue working until they can migrate to using the new file.
+
+## New File Contents
+
+The file will contain the information that was planned for the `.packages` file:
+
+- A *default package* specified by its package name.
+- A list of *packages*, each with:
+  - A *package name*.
+  - A *path* or *location URI*.
+  - A *language version* (major/minor version number)
+
+On top of that, we also add extra metadata which allows us to version the format and attach information about when and how the file was created.
+
+We use JSON for the format. JSON has parsers available for most languages (not all Dart tools are written in Dart) and it is standardized and efficiently parsable.  JSON is *not* easy to write by hand (but doable), and it does not allow comments, but this is a machine generated file that will hardly ever be edited by hand.
+
+## New File Format
+
+The new file will be called `.packages.json` and stored in the same directory as the `.packages` file. The format is JSON text describing is a single JSON object.
+
+That object has, at least, the following properties:
+
+- `apiVersion`: A single integer marking the version of this format, currently `2` (counting `.packages` as the first). Tools should refuse to work on files with a larger version number than the ones they are aware of. Incremental changes will not require increasing the version, so new versions are not expected any time soon, but code should still prepare for it.
```

Comment:
This precludes the possibility of writing 2 formats during a transition period for format upgrades. I worry that future upgrades will be even harder than this one where we will be writing both `.packages` and this file and can smoothly flip over a tool at a time.

Reply:
Interesting idea. 
To have two formats in the same file, we would have to have two separate objects.
Say:
```
{ 
  "v2": {
    ...
  },
  "v3": {
    ...
  }
}
```
and tools should use the most recent version number they are familiar with.

I'm not sure we *want* that, though.

If the new version contains information essential for understanding the program, then the old version plain just doesn't work. Supplying it will not make things work anyway.

If the new version does not contain information that is essential, just useful, then it's an incremental change that doesn't require incrementing the version number to begin with.

So, I can't find a use-case where this would be both necessary and successful.

Reply:
> I can't find a use-case where this would be both necessary and successful.

From my perspective it matches the situation we are in today.

We have many tools that consume the file and few tools that write the file. We were unable to identify and update all the tools that read the file in time to make them compatible with the new version. We chose to go for a smooth rollout and write both versions for some period of time. We specifically chose not to do a hard flip. This new scheme _only_ allows for a hard flip.

How do you envision version changes happening in the future? How will they be different from today such that a "write both" strategy is not necessary?

Reply:
The current change would be a non-breaking in the new format.

The only operation supported by the previous format was to resolve a `package:` URI to a non-`package:` URI. The information needed to do that is available in the new version too, so old code is not broken.

We now add more information, which is used to tell tools how to interpret the files they find.

Future changes are either incremental and non-breaking, or breaking.

If we add more information, but nothing so essential that existing uses of the file will stop working, then it's a non-breaking change. We do not need a new version number.

If we add more information which is necessary for tools to know how files are interpreted, then it's probably a breaking change. Old tools, using only the old information, would no longer be able to correctly interpret the Dart files they find. That's where we need to increment the version, because old code will now do *wrong* things.

Keeping the old information available in the old format, along with new information in a new format, will *not change that*. The old tools are *still broken* because they are simply not capable of handling the new world. I hope that such changes will not happen, or that they will be rare, but I do think that keeping two formats at the same time will *not help*.

That said, we can always have a transitional phase with:
```dart
{
  "apiVersion": 2,
  "forwardCompat": {
    "apiVersion: 3,
    "new data here": ...
  },
  "old data here": ...
}
```
and require version-n readers seeing version m<n data to look for a "forwardCompat" entry which may hold their data. Being a general format allows us such shenanigans. I'd prefer to not decide the exact way to handle this until it becomes necessary, I'll just remember that it is possible if it becomes necessary (and if it actually solves any problem).